providers/redfish/firmware.go

```diff
@@ -0,0 +1,280 @@
+package redfish
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"net/textproto"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	rf "github.com/stmcginnis/gofish/redfish"
+
+	"github.com/bmc-toolbox/bmclib/devices"
+	bmclibErrs "github.com/bmc-toolbox/bmclib/errors"
+)
+
+var (
+	FirmwareApplyAt = []string{
+		devices.FirmwareApplyImmediate,
+		devices.FirmwareApplyOnReset,
+	}
+	FirmwareDownloadProtocols = []string{"HTTP", "NFS", "CIFS", "TFTP", "HTTPS"}
+)
+
+// FirmwareInstall uploads and initiates the firmware install process
+func (c *Conn) FirmwareInstall(ctx context.Context, component, applyAt string, forceInstall bool, reader io.Reader) (jobID string, err error) {
+	// validate firmware update mechanism is supported
+	err = c.firmwareUpdateCompatible(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	// validate applyAt parameter
+	if !stringInSlice(applyAt, FirmwareApplyAt) {
+		return "", errors.Wrap(bmclibErrs.ErrFirmwareInstall, "invalid applyAt parameter: %s"+applyAt)
+	}
+
+	// list redfish firmware install task if theres one present
+	task, err := c.GetFirmwareInstallTaskQueued(ctx, component)
+	if err != nil {
+		return "", err
+	}
+
+	if task != nil {
+		msg := fmt.Sprintf("task for %s firmware install present: %s", component, task.ID)
+		c.Log.V(2).Info("warn", msg)
+
+		if forceInstall {
+			err = c.purgeQueuedFirmwareInstallTask(ctx, component)
+			if err != nil {
+				return "", errors.Wrap(bmclibErrs.ErrFirmwareInstall, err.Error())
+			}
+		} else {
+			return "", errors.Wrap(bmclibErrs.ErrFirmwareInstall, msg)
+		}
+	}
+
+	updateParameters := []byte(
+		fmt.Sprintf(`{"Targets": [], "@Redfish.OperationApplyTime": "%s", "Oem": {}}`, applyAt),
+	)
```

Comment:
Maybe this should be a serialized struct? I don't see any guards for malformed `applyAt` (ex, an `applyAt` with a quote in it) https://go.dev/play/p/4SmSVwhQprM

```suggestion
	updateParameters, err := json.Marshal(struct {
		Targets            []string `json:"Targets"`
		RedfishOpApplyTime string   `json:"@Redfish.OperationApplyTime"`
		Oem                struct{} `json:"Oem"`
	}{
		[]string{},
                applyAt,
		struct{}{},
	})
```

Reply:
thanks!